Source/WebCore/editing/cocoa/AttributedString.mm

```diff
@@ -0,0 +1,160 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+#import "AttributedString.h"
+
+#import <Foundation/Foundation.h>
+#if PLATFORM(MAC)
+#import <AppKit/AppKit.h>
+#else
+#import <pal/ios/UIKitSoftLink.h>
+#endif
+
+namespace WebCore {
+
+bool AttributedString::rangesAreSafe(const String& string, const Vector<std::pair<Range, HashMap<String, AttributeValue>>>& vector)
+{
+    auto stringLength = string.length();
+    for (auto& pair : vector) {
+        auto& range = pair.first;
+        if (range.location > stringLength
+            || range.length > stringLength
+            || range.location + range.length > stringLength)
+            return false;
+    }
+    return true;
+}
+
+static RetainPtr<NSObject> toNSObject(const AttributedString::AttributeValue& value)
+{
+    return WTF::switchOn(value.value, [] (double value) -> RetainPtr<NSObject> {
+        return adoptNS([[NSNumber alloc] initWithDouble:value]);
+    }, [] (const String& value) -> RetainPtr<NSObject> {
+        return (NSString *)value;
+    }, [] (const RetainPtr<NSParagraphStyle>& value) -> RetainPtr<NSObject> {
+        return value;
+    }, [] (const URL& value) -> RetainPtr<NSObject> {
+        return (NSURL *)value;
+#if PLATFORM(MAC)
+    }, [] (const RetainPtr<NSFont>& value) -> RetainPtr<NSObject> {
+        return value;
+    }, [] (const RetainPtr<NSColor>& value) -> RetainPtr<NSObject> {
+        return value;
+    }, [] (const RetainPtr<NSTextAttachment>& value) -> RetainPtr<NSObject> {
+        return value;
+#else
+    }, [] (const RetainPtr<UIFont>& value) -> RetainPtr<NSObject> {
+        return value;
+    }, [] (const RetainPtr<UIColor>& value) -> RetainPtr<NSObject> {
+        return value;
+#endif
+    });
+}
+
+static RetainPtr<NSDictionary> toNSDictionary(const HashMap<String, AttributedString::AttributeValue>& map)
+{
+    auto result = adoptNS([[NSMutableDictionary alloc] initWithCapacity:map.size()]);
```

Comment:
I wonder if this could be initialized more efficiently using `initWithObjects:andKeys`? I guess building the input data for this would be slow because of the need to perform `toNSObject` on each one. But I think Darin created some convenience methods to do bulk conversions somewhere.

Reply:
I don't think that would be much more efficient because we'd have to do additional allocations for the arrays of objects and keys.  I'm already calling initWithCapacity so it shouldn't do more than 1 allocation.